faculty/clients/environment.py

```diff
@@ -30,9 +57,126 @@
         "author_id",
         "created_at",
         "updated_at",
+        "specification",
     ],
 )
 
+EnvironmentUpdate = namedtuple(
+    "EnvironmentUpdate", ["name", "description", "specification"]
+)
+
+
+class SemverSchema(BaseSchema):
+    constraint = EnumField(Constraint, by_value=True, required=True)
+    identifier = fields.String(required=True)
+
+    @post_load
+    def make_version(self, data):
+        return Version(**data)
+
+
+class VersionField(fields.Field):
+    """
+    Field that serialises/deserialises a Python package version.
+    """
+
+    def _deserialize(self, value, attr, obj, **kwargs):
+        if value == "latest":
+            return "latest"
+        else:
+            return SemverSchema().load(value)
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        if value == "latest":
+            return "latest"
+        else:
+            return SemverSchema().dump(value)
+
+
+class PythonPackageSchema(BaseSchema):
+    name = fields.String(required=True)
+    version = VersionField(required=True)
+
+    @post_load
+    def make_python_package(self, data):
+        return PythonPackage(**data)
+
+
+class PipSchema(BaseSchema):
+    extra_index_urls = fields.List(
+        fields.String(), data_key="extraIndexUrls", required=True, many=True
+    )
+    packages = fields.List(fields.Nested(PythonPackageSchema()), required=True)
+
+    @post_load
+    def make_pip(self, data):
+        return Pip(**data)
+
+
+class CondaSchema(BaseSchema):
+    channels = fields.List(fields.String(), required=True)
+    packages = fields.List(fields.Nested(PythonPackageSchema()), required=True)
+
+    @post_load
+    def make_conda(self, data):
+        return Conda(**data)
+
+
+class PythonSpecificationSchema(BaseSchema):
+    conda = fields.Nested(CondaSchema(), required=True)
+    pip = fields.Nested(PipSchema(), required=True)
+
+    @post_load
+    def make_python_specification(self, data):
+        return PythonSpecification(**data)
+
+
+class PythonSchema(BaseSchema):
+    python_2 = fields.Nested(
+        PythonSpecificationSchema(), required=True, data_key="Python2"
```

Comment:
The GUI seemed to implicitly require them - so I thought recreating the GUI expectation was the "safest" thing to do.

Reply:
The GUI always sends them because it was easier for the frontend team to implement it that way.

The backend it not required to send any of `Python2` or `Python3` in the response payload so it will be 'safer' to make them both optional.